src/core/processing/qgsprocessingutils.h

```diff
@@ -317,6 +317,11 @@ class CORE_EXPORT QgsProcessingFeatureSource : public QgsFeatureSource
     QVariant minimumValue( int fieldIndex ) const override;
     QVariant maximumValue( int fieldIndex ) const override;
 
+    /**
+     * Access the underlying original \a source.
+     */
+    QgsFeatureSource *source() const;
```

Comment:
Ok - here's the bit I'm 50-50 on. I really want to ensure that algorithms never try and abuse feature sources by e.g. calling methods which are available only in feature source subclasses (e.g. calling a vector layer method). You'll notice that I haven't added any python sip subclassing for QgsFeatureSource for this very reason.

So - just to keep things safe, can you make this getter private and friend class QgsProcessingAlgorithm? Either that or block it from the python api.

Reply:
Alternatively, there could also be a `QgsExpressionContext expressionContext() const` method here, doing the cast to `QgsExpressionContextGenerator` in here.

Reply:
Yep, that would also work and is nicer!

Reply:
Thinking about this again.

Isn't it possible that in another context someone wants to do something similar. I.e. an algorithm requires at least a source, but - if available - it will make use of additional vector layer properties. I'm not sure it's so bad to expose the underlying QgsFeatureSource.
What's the reason to hide it away?

Reply:
Because I don't want algorithms to become reliant on this behaviour, and expect a QgsVectorLayer when they declare a feature source parameter. In future I hope to make greater use of non vector-layer feature sources in processing - e.g. if we move to a feature "pipeline" approach, where the feature source is actually just an iterator coming from the previous model step.

If algorithms NEED access to a vector layer and all the corresponding api, they should declare the parameter as a QgsProcessingParameterVectorLayer  instead of a QgsProcessingParameterFeatureSource. Then they'll have access to the full layer api,  but will also miss out on all the potential future improvements available only to feature sources.

For now, the main benefit to declaring parameters as feature source parameters is that they allow the "selected features only" option. QgsProcessingParameterVectorLayer parameters don't get this option.

Reply:
(Another thought I have is of making use of QgsVectorLayerCache inside a model when an underlying layer is used by multiple algorithms OR iterated multiple times within a single algorithm. It should be easy to make QgsVectorLayerCache expose the QgsFeatureSource interface, and then algorithms which only depend on QgsFeatureSource API will work without issue... but if they depend on the whole QgsVectorLayer API, then no luck!)

Reply:
I don't see this risk that high. It all depends on the available examples and the API docs.

Most processing algorithms will just copy-paste existing ones, so if the existing ones use a QgsFeatureSource, so will the new ones. I.e. let's make sure we don't do the QgsVectorLayer vs. QgsVectorDataProvider mistake again, but let's not cripple the API just for this.

Reply:
About the caching of features in a pipe scenario, I was thinking to put `QQueue`s between the source and the sinks, no random access will be required in this scenario and this should keep the memory footprint reasonably low (at least if the sink consumes faster than the source produces).